src/twisted/conch/ssh/keys.py

```diff
@@ -921,15 +921,21 @@ def _fromEd25519Components(cls, a, k=None):
 
         return cls(keyObject)
 
-    def __init__(self, keyObject):
+    def __init__(self, keyObject, hashAlgorithm=None):
         """
         Initialize with a private or public
         C{cryptography.hazmat.primitives.asymmetric} key.
 
         @param keyObject: Low level key.
         @type keyObject: C{cryptography.hazmat.primitives.asymmetric} key.
+
+        @param hashAlgorithm: The hash algorithm to use when signing and
+        verifying using this key, or L{None} to use a default appropriate
+        for the key type.
+        @type hashAlgorithm: L{hashes.HashAlgorithm} or L{None}
         """
         self._keyObject = keyObject
+        self.hashAlgorithm = hashAlgorithm
```

Comment:
Do we need it as public writable member?

I would prefer to have it as a read-only property.

I think that hashAlgorithm is something that is defined once and then is only used internally.

I am hoping that by making it private and only setting it at construct time, we can prevent  some future strange bugs in which the hashAlgoritm is updated at runtime.

Reply:
This comment caused me to realize that this whole approach was fundamentally based on woolly thinking.  The public key signature algorithm (and hence the hash algorithm) is fundamentally not a property of the key; it's a property of how you use the key, by signing or verification.  When I went and looked at how the client equivalent of this work might work in future, I found that we'll really need to keep the concepts more clearly separated there, since a client picks a key and then picks an appropriate key signature algorithm for it; having to construct a copy of the key using a different hash algorithm would be very cumbersome.  So I'm glad you saved me from locking us into an inappropriate API!

Instead of all this, I've changed `Key.sign` to take an optional `signatureType` argument.  This turns out to be quite straightforward, and in particular it means there's no messing about in the factory to make variant copies of keys (although I did have to teach the factory to fill in `supportedPublicKeys` properly).  The patch is a bit longer, but I think it's now conceptually clearer.